Add install target.

--- cctz-v2.2/CMakeLists.txt	2018-04-08 21:06:53.000000000 +0200
+++ cctz-v2.2/CMakeLists.txt	2018-04-08 21:12:51.000000000 +0200
@@ -138,3 +138,20 @@ if (BUILD_TESTING)
 endif()
 
 feature_summary(WHAT ALL FATAL_ON_MISSING_REQUIRED_PACKAGES)
+
+install(TARGETS cctz 
+        LIBRARY DESTINATION lib
+        ARCHIVE DESTINATION lib
+        RUNTIME DESTINATION bin
+        INCLUDES DESTINATION include)
+
+install(
+      DIRECTORY
+        include/cctz
+      DESTINATION
+        include
+      COMPONENT
+        Devel
+    )
+    
+ set_target_properties(cctz PROPERTIES VERSION 2.2.0 SOVERSION 2.2.0)

